### ISSUE TITLE
sapply

### ISSUE THREAD

**Issue by MaggiK** [pearselab/r-intro-MaggiK#4]
I am still having problems wrapping my head around sapply. For number four I have:

magnificent_summary_func<- function(x){
sapply(x, 
  if(is.numeric(x[i,])){
    return(my_summary_func(x[i,]))
    } else {
      return(my_categ_summary_func(x)[i,])
    } 
)
}

The goal is to use sapply to iterate through the matrix and is is.numeric where numeric and the others get the categorical summary. The way I have it doesn't work. It doesn't know what i is. I am not sure how to continue or if I am way off. Thanks.

**Comment by willpearse** [comment]
In this exercise, I'm trying to teach you to use the `apply` family, which means you won't be iterating on anything. Indeed, you haven't typed `for` anywhere in this, so R probably doesn't even know what `i` is as you haven't defined it in a loop :-(

Try to break it down into the following steps:

```
# Create vector of which columns are numeric (using sapply)
# Call numeric function on subset of data.frame that is numeric
# Call non-numeric function on subset of data.frame that is not numeric
# Voila!
```

If you follow the steps above, the whole function will be three lines long, and won't involve an `if` anywhere. Remember the subsetting we talked about during our session: you're not doing that in your code snippet, and if you do subset it will be much easier to write

**Comment by MaggiK** [comment]
So something like this?

numer<- sapply(vec, is.numeric)
magnificent_summary_func<- function(x){
  my_summary_func(numer)
  my_categ_summary_func(x)
}

This doesn't seem to separate a data frame. In this numer comes up as FALSE for each value of a mixed dataframe. Do I need to add something to the sapply argument?

**Comment by willpearse** [comment]
Yes. You need to subset the data.frame, and you need to do it inside your function. Give it another try, and show me what you come up with.

**Comment by MaggiK** [comment]
ok. Added a subset to the function. I think this should work. But my problem may be the dataframe I am testing it with. 

numer<- sapply(bn, is.numeric)
magnificent_summary_func<- function(x){
  nm_x<-subset(x, numer)
  my_summary_func(nm_x)
  my_categ_summary_func(x)
}

**Comment by willpearse** [comment]
OK, you can do subsetting using the `subset` function (or with the `[]`) - that's great!

The problem you have now, by the look of it, is you've pulled out the numeric ones. How are you going to pull out the ones that are NOT numeric? Hint: what is the R operator for NOT?...
